wc: compute minimum width to format counts up front

jfinkels opened the pull request:
This pull request fixes two issues with the string formatting width for counts displayed by `wc`.

First, the output was previously not using the default minimum width (seven characters) when reading from `stdin`. This commit corrects this behavior to match GNU `wc`. For example,

    $ cat alice_in_wonderland.txt | wc
          5      57     302

Second, if at least 10^7 bytes were read from `stdin` *after* reading from a smaller regular file, then every output row would have width 8. This disagrees with GNU `wc`, in which only the `stdin` row and the total row would have width 8. This commit corrects this behavior to match GNU `wc`. For example,

    $ printf "%.0s0" {1..10000000} | wc emptyfile.txt -
	  0       0       0 emptyfile.txt
	  0       1 10000000
	  0       1 10000000 total

(I did not add a test for this last case because I thought it might make the tests too slow. Let me know if it makes sense to add a test.)

Fixes #2186.

jfinkels commented:
This change has an additional benefit that is not obvious from this pull request. GNU `wc` prints rows of output as they are computed, but formatted with the appropriate width even if the largest file appears at the end of the list of files. This pull request facilitates making that change as well. (Currently, the `uutils` version of `wc` waits until all inputs have been processed before printing any rows of output.)

jfinkels commented on src/uu/wc/src/wc.rs:
```diff
@@ -214,11 +216,70 @@ fn word_count_from_path(path: &str, settings: &Settings) -> WcResult<WordCount>
+        if !metadata.is_file() {
+            continue;
+        }
+        let num_bytes = metadata.len();
```
This has the downside of computing the number of bytes in a file twice, once here and once during the processing of files. A future pull request can eliminate that redundancy.

sylvestre replied:
can you add a comment with TODO and create an issue about this ? thanks

jfinkels replied:
Yes, I will add a comment and create a new GitHub issue.

jfinkels replied:
The issue is #2201.

jfinkels commented:
I have resolved the conflicts, rebased, and squashed all commits into one. I slightly changed the implementation (created one more helper function) so please give it another quick look.

sylvestre commented on src/uu/wc/src/wc.rs:
```diff
@@ -254,11 +256,101 @@ fn word_count_from_input(input: &Input, settings: &Settings) -> WcResult<WordCou
+    let mut error_count = 0;
+    let max_width = max_width(&inputs).unwrap_or_else(|err| {
+        show_error!("{}", err);
+        error_count += 1;
```
would it be possible to add a test to go in this path ?

jfinkels replied:
Ah, I see the warnings produced by codecov. Yes, I will try to come up with a test case that covers this path.

jfinkels commented:
I have updated the code slightly one more time to cover untested code paths. That revealed some incompatibilities with GNU `wc` in the case of a nonexistent file and attempting to read from a directory. I have added test cases for those two situations with comments describing the incompatibilities. I can open new issues as well to describe what needs to be done to make the code compatible with GNU `wc`.

sylvestre commented:
Seems that the Windows test needs to be refreshed
```

---- test_wc::test_read_from_nonexistent_file stdout ----
current_directory_resolved: 
run: D:\a\coreutils\coreutils\target\x86_64-pc-windows-msvc\debug\coreutils.exe wc bogusfile
thread 'test_wc::test_read_from_nonexistent_file' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<"wc: error: The system cannot find the file specified. (os error 2)"
>"wc: error: No such file or directory (os error 2)"
```

jhscheer commented on tests/by-util/test_wc.rs:
```diff
@@ -168,3 +170,58 @@ fn test_file_one_long_word() {
+    //
+    new_ucmd!()
+        .args(&["bogusfile"])
+        .run()
```
If the test is expected to fail, please use `fails()` instead of `run()` (checks if error code is 1).
For the error messages I would just test like this: `stderr_contains("bogusfile: No such file or directory")`.
Because the prefix is subject of change if/when the macros are changes.

jfinkels replied:
Thanks for the advice, I will make those changes.

jfinkels commented:
> Seems that the Windows test needs to be refreshed

Hmm perhaps it would be simpler to fix issue #2211 first, in order to make the error messages uniform on all platforms?

jfinkels commented:
I have rebased on master, now that #2211 is resolved. I don't have a Windows computer, so I hope those changes will help resolve the tests.
